packages/blunderbuss/src/datastore-lock.ts

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {v4 as uuidv4} from 'uuid';
+import {Datastore, Key} from '@google-cloud/datastore';
+import {logger} from 'gcf-utils';
+
+const DEFAULT_LOCK_EXPIRY = 20 * 1000; // 20 seconds
+const LOCK_ACQUIRE_TIMEOUT = 120 * 1000; // 120 seconds
+const BACKOFF_INITIAL_DELAY = 2 * 1000; // 1 seconds
+const BACKOFF_MAX_DELAY = 10 * 1000; // 10 seconds
+
+enum AcquireResult {
+  Success = 1,
+  Failure,
+  LockActive,
```

Comment:
I'm not sure I understand the differentiation between Failure / LockActive. It seems like the possible outcomes are:

1. Acquired the lock successfully (Success)
2. The lock is acquired by someone else (Failure)
3. Something else unexpected happens

It seems like #3 should just throw an error?

Reply:
Datastore can throw an error when there are too many writes on one single entity. In that case, we want to retry with some delay.

Reply:
The logic for both sections is [essentially the same](https://github.com/googleapis/repo-automation-bots/pull/1732/files#diff-3215fb0ab8076b6916f2a23222a8360fe81a7cafdbc9e2813c73e7e6012792d4R79-R88)

Would it be better to just throw an error when the lock is already taken and then just retry on any error? 

At some point you might want to introduce functionality for providing a timeout for acquiring the lock, in which case returning the error from the most recent attempt that doesn't exceed the user provided timeout will likely be useful.

Reply:
Sorry I don't understand. If the lock is active, wait and retry is the most natural thing to do...

Reply:
It's more of a nit at this point since they are handled the same, but what I'm trying to say is:

1. I agree with you that retrying on an error makes sense
2. I agree with you that retrying on lock being taken makes sense. However, since we are already retrying on an error because of the above statement, there is no reason to treat already locked locks special. Treat it as an error that is retried with the same logic. 
3.  Finally, I think that the above is more desirable behavior because (probably) eventually, user's of your lock may want to timeout how long they wait for it (something like `await lock.acquire(30s)`). In these situations, it would be better to not return true/false, but instead have no return value and assume that if it completed if it locked, and if it's not it'll throw an error (which will then contain the root cause - it was already locked, database errors, etc).

Reply:
> 2
I understand. It's more of a sake of development. It is better to emit different debug message during development.

> 3
Currently only failure is by timeout.